### PR TITLE
stage1: unicode characters in character literals and utf-8 validation

### PR DESCRIPTION
**Pull request by shawnl**
adding utf-8 validation revealed some non-utf-8 stuff in std

Closes: #2097 

I'm looking at the self-hosted compiler now....

### CODE REVIEW

File: std/zig/tokenizer.zig [hunk 1]
```diff
@@ -673,6 +687,9 @@ pub const Tokenizer = struct {
                         seen_escape_digits = 0;
                         expected_escape_digits = 6;
                     },
+                    'n', 'r', '\\', 't', '\'', '\"' => {
+                        // TODO
```
Comment (andrewrk) [inline std/zig/tokenizer.zig]: Can you make this `@panic("TODO");`? Better to crash here than to silently be incorrect.

File: std/unicode.zig [hunk 2]
```diff
@@ -18,11 +18,25 @@ pub fn utf8CodepointSequenceLength(c: u32) !u3 {
...
-    return error.Utf8InvalidStartByte;
+    var ret: u4 = undefined;
+    ret = utf8ByteSequenceLengthNoValidate(first_byte);
+    if (ret > 4)
+        return error.Utf8InvalidStartByte;
+    return @intCast(u3, ret);
+}
+
+pub fn utf8ByteSequenceLengthNoValidate(first_byte: u8) u4 {
+    const utf8_skip_data = []u8{
```
Comment (andrewrk) [inline std/unicode.zig]: Can you show evidence this is faster than the bit manipulations from before? It's quite possible the memory lookups are slower than the bit manipulation instructions. and when there is no perf evidence, the clearer code is the winner, and I think it's fair to say the bit manipulation code here is clearer.

In general it's reasonable to have `utf8ByteSequenceLength` return a possible error, and then tokenizer.zig can use `catch unreachable` if it can guarantee there will be no error.

File: std/pdb.zig [hunk 3]
```diff
@@ -558,7 +558,7 @@ const SuperBlock = packed struct {
     BlockSize: u32,
 
     /// The index of a block within the file, at which begins a bitfield representing
-    /// the set of all blocks within the file which are “free” (i.e. the data within
```
Comment (thejoshwolfe) [inline std/pdb.zig]: these changes are an improvement independent of this PR. you could make these a separate PR.

File: std/unicode.zig [hunk 4]
```diff
@@ -17,12 +17,29 @@ pub fn utf8CodepointSequenceLength(c: u32) !u3 {
...
+      1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+      1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+      1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+      1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+      1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+      1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+      2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
+      3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,5,5,5,5,6,6,7,8, // Using \xfe and \xff allows utf-8 to encode 42-bits
```
Comment (thejoshwolfe) [inline std/unicode.zig]: As i mentioned above, this is not the correct definition of UTF-8.

File: std/zig/tokenizer.zig [hunk 5]
```diff
@@ -235,6 +236,7 @@ pub const Tokenizer = struct {
         StringLiteralBackslash,
         MultilineStringLiteralLine,
         CharLiteral,
+        CharLiteralUTF8,
```
Comment (thejoshwolfe) [inline std/zig/tokenizer.zig]: is this state ever entered? (is this code unfinished?)

File: std/unicode.zig [hunk 6]
```diff
@@ -146,21 +166,12 @@ pub fn utf8Decode4(bytes: []const u8) Utf8Decode4Error!u32 {
     return value;
 }
 
+// TODO replace with something faster
```
Comment (thejoshwolfe) [inline std/unicode.zig]: this is unhelpful.

Comment (thejoshwolfe) [reply std/unicode.zig]: sorry, that came off as rude. i mean it would be better to link to some faster implementations, like you did in the C++ code, or maybe just say "optimize with SIMD" or some suggestion to make it faster instead of this vague comment.

File: std/unicode.zig (hunk 2 above)
Comment (shawnl) [reply std/unicode.zig]: I developed a nice branchless version with `@clz` that I think you will like

File: std/pdb.zig (hunk 3 above)
Comment (shawnl) [reply std/pdb.zig]: This PR doesn't pass without fixing all the broken code. This code had non-UTF8.

File: std/unicode.zig (hunk 4 above)
Comment (shawnl) [reply std/unicode.zig]: It is fine, as long as you test for len > 4 (utf-8 can encode up to 42-bits, but that is not Unicode, they are two differn't standards) But I wrote a much more obvious version with `@clz()` that I think you will like.`

Comment (shawnl) [reply std/unicode.zig]: The skip data comes from a totally differn't code set, because while the validate code doesn't need it, I need this information.

**Comment by daurnimator** [comment]
Note that most people don't want utf-8, but [wtf-8](https://simonsapin.github.io/wtf-8/) which is what windows uses for file paths, and what you encode in e.g. json.

**Comment by thejoshwolfe** [comment]
@daurnimator I don't think that's right. The link you provided says that Windows and JavaScript don't enforce that UTF-16 surrogate halves are in pairs. This has more to do with transforming between  UTF-16's and UTF-8, which is not happening in this PR. That's important for the Windows os code, but not for UTF-8 code. (And I don't think the JavaScript nonconformance is relevant to Zig. It's definitely not relevant to conformant [JSON](https://www.json.org), which is specified in terms of "Unicode characters", not bytes or UTF-16.)

File: std/zig/tokenizer.zig (hunk 5 above)
Comment (shawnl) [reply std/zig/tokenizer.zig]: Yes.

**Comment by shawnl** [comment]
I feel this review got derailed by reading some code that was not meant to validate the utf-8 as validation code.....That code was just to get the length of a code-point, independent of validation, and it did that correctly.

**Comment by daurnimator** [comment]
> I really don't think this is true. There was a version of UTF-8 that could encode more than 24 bits, but that version of UTF-8 was obsoleted in 2003. 

That is correct.

> And my research into the standardization of UTF-8 says that it is tightly coupled with the standardization of Unicode. Perhaps you can point me to some documentation that would clear this up. I'm getting my info from mostly from https://en.wikipedia.org/wiki/UTF-8 .

Not particularly. UTF-8 was set in stone as [rfc3629](https://tools.ietf.org/html/rfc3629)/[std63](https://www.rfc-editor.org/std/std63.txt) which was written in in 2003.

File: std/unicode.zig [hunk 7]
```diff
@@ -17,12 +17,30 @@ pub fn utf8CodepointSequenceLength(c: u32) !u3 {
...
+    var clz: u4 = undefined;
+    const INVALID = u8(9);
+    clz = @clz(~first_byte);
+    len = switch (clz) {
+        0 => u8(1), // ASCII
+        1 => INVALID, // Continuation byte
+        2 => u8(2),
+        3 => u8(3),
+        4 => u8(4),
+        5 => u8(5),
```
[the remark this answers is not in the document]
Comment (shawnl) [reply std/unicode.zig]: But that way llvm can't optimize the switch into a look-up table. I was specifically targeting that llvm optimization:

https://zig.godbolt.org/z/c1AQfs

**Comment by shawnl** [comment]
@thejoshwolfe It was a latin-1 curved quote character. You have to look at the original because github coverts it into UTF-8.

OK, I will work on a nieve C99 implementation of UTF-8 validation that you find is clear. I actually thought this version was clearer because it is a strict finite-state-automata........

**Comment by thejoshwolfe** [comment]
I was looking at the original, or at least what i got from git. See the Python 2 code that renders the bytes if the file? That's utf8. Did your file get transformed somehow? or are we looking at different git commits maybe? I don't think git would convert between charsets automatically without some setting, and i didn't enable anything like that.

File: std/unicode.zig (hunk 7 above)
[the remark this answers is not in the document]
Comment (thejoshwolfe) [reply std/unicode.zig]: does this work? (godbolt doesn't work on mobile, and i can't really read x86 assembly anyway.)

```zig
    const leading_ones = @clz(~first_byte);
    if (leading_ones == 1 or leading_ones > 4) return -1;
    return leading_ones;
```

**Comment by shawnl** [comment]
Your right. My bad. I guess I was using a buggy early version of my code, and didn't check if it was actually latin-1 which I suspected.

File: std/unicode.zig (hunk 7 above)
[the remark this answers is not in the document]
Comment (shawnl) [reply std/unicode.zig]: That be two branches anyways (== 1 and > 4). My code uses only one branch.

File: std/unicode.zig [hunk 8]
```diff
@@ -461,7 +475,10 @@ fn testDecode(bytes: []const u8) !u32 {
     const length = try utf8ByteSequenceLength(bytes[0]);
     if (bytes.len < length) return error.UnexpectedEof;
     testing.expect(bytes.len == length);
-    return utf8Decode(bytes);
+    var c: u32 = undefined;
+    c = 42; // Shut up compiler
```
[the remark this answers is not in the document]
Comment (shawnl) [reply std/unicode.zig]: Thank you. I was missing a `try`.

File: std/unicode.zig (hunk 7 above)
[the remark this answers is not in the document]
Comment (andrewrk) [reply std/unicode.zig]: Write the code to express intent. If you want to write code that deviates from that, you have to measure performance, not just look at the assembly. Here's a benchmark:

```zig
const std = @import("std");

const Utf8ByteSequenceLengthError = error{Invalid};

pub fn utf8ByteSequenceLengthOptimized(first_byte: u8) Utf8ByteSequenceLengthError!u3 {
    var clz = @clz(~first_byte);
    var len = switch (clz) {
        0 => u8(1), // ASCII
        1 => u8(9), // Continuation byte
        2 => u8(2),
        3 => u8(3),
        4 => u8(4),
        5 => u8(5),
        6 => u8(6),
        7 => u8(7),
        8 => u8(8),
        else => unreachable,
    };
    if (len > 4) {
        return error.Invalid;
    }
    return @intCast(u3, len);
}

pub fn utf8ByteSequenceLengthSimple(first_byte: u8) Utf8ByteSequenceLengthError!u3 {
    switch (@clz(~first_byte)) {
        0 => return 1, // ASCII
        1 => return error.Invalid,
        2 => return 2,
        3 => return 3,
        4 => return 4,
        else => return error.Invalid,
    }
}

pub fn main() !void {
    const Item = struct {
        func: fn (num: u8) Utf8ByteSequenceLengthError!u3,
        name: []const u8,
    };
    const items = []Item{
        Item{
            .func = utf8ByteSequenceLengthOptimized,
            .name = "optimized",
        },
        Item{
            .func = utf8ByteSequenceLengthSimple,
            .name = "simple",
        },
        Item{
            .func = utf8ByteSequenceLengthOptimized,
            .name = "optimized",
        },
        Item{
            .func = utf8ByteSequenceLengthSimple,
            .name = "simple",
        },
    };
    const seed = 0x1234;
    // first make sure the values are all equal
    {
        var prng = std.rand.DefaultPrng.init(seed);
        var i: usize = 0;
        var sum: usize = 0;
        while (i < 10000000) : (i += 1) {
            const rand_byte = prng.random.int(u8);
            const opt_result = if (utf8ByteSequenceLengthOptimized(rand_byte)) |x| usize(x) else |_| 100;
            const simple_result = if (utf8ByteSequenceLengthSimple(rand_byte)) |x| usize(x) else |_| 100;
            if (opt_result != simple_result) {
                std.debug.panic("bad value: {}\n", rand_byte);
            }
        }
    }
    for (items) |item| {
        var prng = std.rand.DefaultPrng.init(seed);
        var timer = try std.os.time.Timer.start();
        var i: usize = 0;
        var sum: usize = 0;
        const start = timer.lap();
        while (i < 10000000) : (i += 1) {
            const rand_byte = prng.random.int(u8);
            if (item.func(rand_byte)) |x| {
                sum += x;
            } else |_| {
                sum += 100;
            }
        }
        const end = timer.read();
        const elapsed_s = @intToFloat(f64, end - start) / std.os.time.ns_per_s;
        const bytes_per_sec = @floatToInt(usize, @intToFloat(f64, i) / elapsed_s);
        std.debug.warn("{}: {Bi1}/sec, checksum: {}\n", item.name, bytes_per_sec, sum);
    }
}
```

```
$ ./zig build-exe test.zig 
$ ./test
optimized: 8.0MiB/sec, checksum: 291975483
simple: 8.9MiB/sec, checksum: 291975483
optimized: 8.0MiB/sec, checksum: 291975483
simple: 8.9MiB/sec, checksum: 291975483
$ ./zig build-exe test.zig --release-fast
$ ./test
optimized: 69.2MiB/sec, checksum: 291975483
simple: 69.3MiB/sec, checksum: 291975483
optimized: 69.4MiB/sec, checksum: 291975483
simple: 69.5MiB/sec, checksum: 291975483
```

So in debug mode the simple version is marginally faster, and in release-fast mode, they're indistinguishable.

Write code to express intent.

Even if this benchmark had produced results showing clearly that the non-simple one was faster, it would be better to put effort towards improving LLVM optimization passes to recognize a pattern, rather than to modify zig code to work around the implementation details of the optimizer.

[the remark this answers is not in the document]
Comment (shawnl) [reply std/unicode.zig]: The benchmark is significant if you take the PRNG out of the loop:
```
optimized: 175.2MiB/sec, checksum: 291975483
simple: 124.9MiB/sec, checksum: 291975483
optimized: 171.3MiB/sec, checksum: 291975483
simple: 126.9MiB/sec, checksum: 291975483

const std = @import("std");
const Utf8ByteSequenceLengthError = error{
    Utf8InvalidStartByte,
};
pub fn utf8ByteSequenceLengthOptimized(first_byte: u8) Utf8ByteSequenceLengthError!u3 {
    const INVALID = 0;
    const swtch = []u8{1, INVALID, 2, 3, 4, INVALID, INVALID, INVALID, INVALID};
    var len = swtch[@clz(~first_byte)];
    if (len == INVALID) {
        return error.Utf8InvalidStartByte;
    }
    return @intCast(u3, len);
}

pub fn utf8ByteSequenceLengthSimple(first_byte: u8) Utf8ByteSequenceLengthError!u3 {
    if (first_byte < 0b10000000) return u3(1);
    if (first_byte & 0b11100000 == 0b11000000) return u3(2);
    if (first_byte & 0b11110000 == 0b11100000) return u3(3);
    if (first_byte & 0b11111000 == 0b11110000) return u3(4);
    return error.Utf8InvalidStartByte;
}

const Allocator = std.mem.Allocator;

pub fn main() !void {
    const Item = struct {
        func: fn (num: u8) Utf8ByteSequenceLengthError!u3,
        name: []const u8,
    };
    const items = []Item{
        Item{
            .func = utf8ByteSequenceLengthOptimized,
            .name = "optimized",
        },
        Item{
            .func = utf8ByteSequenceLengthSimple,
            .name = "simple",
        },
        Item{
            .func = utf8ByteSequenceLengthOptimized,
            .name = "optimized",
        },
        Item{
            .func = utf8ByteSequenceLengthSimple,
            .name = "simple",
        },
    };
    const seed = 0x1234;
    const allocator = &std.heap.DirectAllocator.init().allocator;
    const d = try allocator.alloc(u8, 10000000);
    {
        var prng = std.rand.DefaultPrng.init(seed);
        var i: usize = 0;
        var sum: usize = 0;
        while (i < 10000000) : (i += 1) {
            d[i] = prng.random.int(u8);
        }
    }
    // first make sure the values are all equal
    {
        var i: usize = 0;
        var sum: usize = 0;
        while (i < 10000000) : (i += 1) {
            const opt_result = if (utf8ByteSequenceLengthOptimized(d[i])) |x| usize(x) else |_| 100;
            const simple_result = if (utf8ByteSequenceLengthSimple(d[i])) |x| usize(x) else |_| 100;
            if (opt_result != simple_result) {
                std.debug.panic("bad value: {}\n", d[i]);
            }
        }
    }
    for (items) |item| {
        var timer = try std.os.time.Timer.start();
        var i: usize = 0;
        var sum: usize = 0;
        const start = timer.lap();
        while (i < 10000000) : (i += 1) {
            if (item.func(d[i])) |x| {
                sum += x;
            } else |_| {
                sum += 100;
            }
        }
        const end = timer.read();
        const elapsed_s = @intToFloat(f64, end - start) / std.os.time.ns_per_s;
        const bytes_per_sec = @floatToInt(usize, @intToFloat(f64, i) / elapsed_s);
        std.debug.warn("{}: {Bi1}/sec, checksum: {}\n", item.name, bytes_per_sec, sum);
    }
}
```

[the remark this answers is not in the document]
Comment (andrewrk) [reply std/unicode.zig]: OK great. Definitely we want the `utf8ByteSequenceLengthOptimized` that you have here.

For what it's worth, I actually think that both of these functions are equally readable:

```zig
pub fn utf8ByteSequenceLengthOptimized(first_byte: u8) Utf8ByteSequenceLengthError!u3 {
    const INVALID = 0;
    const swtch = []u8{1, INVALID, 2, 3, 4, INVALID, INVALID, INVALID, INVALID};
    var len = swtch[@clz(~first_byte)];
    if (len == INVALID) {
        return error.Utf8InvalidStartByte;
    }
    return @intCast(u3, len);
}

pub fn utf8ByteSequenceLengthSimple(first_byte: u8) Utf8ByteSequenceLengthError!u3 {
    if (first_byte < 0b10000000) return u3(1);
    if (first_byte & 0b11100000 == 0b11000000) return u3(2);
    if (first_byte & 0b11110000 == 0b11100000) return u3(3);
    if (first_byte & 0b11111000 == 0b11110000) return u3(4);
    return error.Utf8InvalidStartByte;
}
```

**Comment by andrewrk** [comment]
This is also too much. One pull request at a time from you please, which only does 1 thing. I can't keep up with this. You removed shebangs in this PR and that's not even mentioned in the pull request description.
